File: packages/core/src/parser/__testdata__/sample.kt
package com.synapse.core

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.util.concurrent.ConcurrentHashMap

const val MAX_RETRIES = 3
const val DEFAULT_MODEL = "claude-opus-4-6"

enum class Status { ACTIVE, IDLE, STOPPED, ERROR }

data class AgentConfig(
    val model: String = DEFAULT_MODEL,
    val maxTokens: Int = 4096,
    val temperature: Double = 0.7,
)

sealed class AgentResult {
    data class Success(val message: String) : AgentResult()
    data class Error(val code: Int, val message: String) : AgentResult()
}

interface Agent {
    suspend fun process(message: String): AgentResult
    fun getTools(): List<String>
    val status: Status
}

interface Configurable {
    fun configure(config: AgentConfig)
}

abstract class BaseAgent(val name: String) : Agent {
    protected var _status: Status = Status.IDLE
    override val status get() = _status

    protected fun validate(input: String): Boolean = input.isNotBlank()
}

class SynapseAgent(
    name: String,
    private var config: AgentConfig = AgentConfig(),
) : BaseAgent(name), Configurable {

    override suspend fun process(message: String): AgentResult =
        withContext(Dispatchers.IO) {
            _status = Status.ACTIVE
            val result = callModel(message)
            _status = Status.IDLE
            AgentResult.Success(result)
        }

    override fun getTools(): List<String> = listOf("search", "read", "write")

    override fun configure(config: AgentConfig) {
        this.config = config
    }

    private suspend fun callModel(message: String): String {
        // TODO: implement actual API call
        return "Response to: $message"
    }

    companion object {
        fun create(name: String) = SynapseAgent(name)
    }
}

object AgentRegistry {
    private val agents = ConcurrentHashMap<String, Agent>()

    fun register(agent: SynapseAgent) { agents[agent.name] = agent }
    fun get(name: String): Agent? = agents[name]
    fun listAll(): List<Agent> = agents.values.toList()
}

fun createAgent(name: String, config: AgentConfig? = null): SynapseAgent =
    SynapseAgent(name, config ?: AgentConfig())

// FIXME: add proper error handling in process

File: packages/core/src/parser/__testdata__/sample.php
<?php

namespace Synapse\Core;

use Synapse\Config\AgentConfig;
use Psr\Log\LoggerInterface;

define('MAX_RETRIES', 3);
const DEFAULT_MODEL = 'claude-opus-4-6';

enum Status: string {
    case Active = 'active';
    case Idle = 'idle';
    case Stopped = 'stopped';
    case Error = 'error';
}

interface AgentInterface {
    public function process(string $message): string;
    public function getTools(): array;
    public function getStatus(): Status;
}

interface Configurable {
    public function configure(AgentConfig $config): void;
}

trait HasLogging {
    private LoggerInterface $logger;

    protected function log(string $message): void {
        $this->logger->info("[{$this->name}] {$message}");
    }
}

abstract class BaseAgent implements AgentInterface {
    use HasLogging;

    protected Status $status = Status::Idle;
    public const VERSION = '1.0.0';

    public function __construct(
        protected readonly string $name,
        protected AgentConfig $config,
    ) {}

    abstract public function process(string $message): string;

    protected function validate(string $input): bool {
        return !empty(trim($input));
    }
}

class SynapseAgent extends BaseAgent implements Configurable {
    public function process(string $message): string {
        $this->status = Status::Active;
        $result = $this->callModel($message);
        $this->status = Status::Idle;
        return $result;
    }

    public function getTools(): array {
        return ['search', 'read', 'write'];
    }

    public function getStatus(): Status {
        return $this->status;
    }

    public function configure(AgentConfig $config): void {
        $this->config = $config;
    }

    private function callModel(string $message): string {
        // TODO: implement actual model call
        return "Response to: {$message}";
    }

    public static function create(string $name): self {
        return new self($name, new AgentConfig());
    }
}

function createAgent(string $name, ?AgentConfig $config = null): SynapseAgent {
    return new SynapseAgent($name, $config ?? new AgentConfig());
}

// FIXME: add retry logic for model calls
